lib/core/src/irods_network_manager.cpp

```diff
@@ -1,118 +1,196 @@
+
+            // check table again in case the plugin was added while we were waiting on the lock
+            auto plugin_itr_ = plugins_.find(key);
+            if (plugin_itr_ != plugins_.end()) {
+                _net = plugin_itr_->second;
+                return SUCCESS();
+            }
+
+            std::string type = _type;
+            type += (CLIENT_PT == _proc_type) ? "_client" : "_server";
+
+            // load plugin
+            network_ptr ptr;
+            error ret = load_network_plugin(ptr, type, _inst, _ctx);
+            if (!ret.ok()) {
+                return PASSMSG("Failed to load network plugin", ret);
+            }
+
+            // insert plugin into table
+            bool was_inserted = false;
+            std::tie(plugin_itr_, was_inserted) = plugins_.try_emplace(key, ptr);
+            if (!was_inserted) {
+                return ERROR(SYS_INTERNAL_ERR, "Plugin already in table despite lock");
+            }
+
+            _net = plugin_itr_->second;
+
+            // insert into legacy plugins table (for deprecated functions)
+            ret = plugins_old_.set(_inst, ptr);
+            if (!ret.ok()) {
+                log(LOG_WARNING, ret.result());
+            }
+            else {
+                // verify insertion into legacy table
+                network_ptr temp_ptr;
+                ret = plugins_old_.get(_inst, temp_ptr);
+                if (!ret.ok()) {
+                    log(LOG_WARNING, ret.result());
+                }
+                else if (_net != temp_ptr) {
+                    log(LOG_WARNING, "legacy plugins table doesn't match regular plugins table");
+                }
+            }
+        }
+
+        return SUCCESS();
+    }
+
+    // =-=-=-=-=-=-=-
+    // deprecated - retrieve a network plugin given its key
+    error network_manager::resolve(const std::string& _key, network_ptr& _value)
+    {
+        if (_key.empty()) {
+            return ERROR(SYS_INVALID_INPUT_PARAM, "empty key");
+        }
+
+        // we're going to read the plugins table
+        // acquire shared lock so writers block
+        std::shared_lock lock(table_lock_);
+        auto plugin_itr_ = plugins_old_.find(_key);
+
+        if (plugin_itr_ == plugins_old_.end()) {
+            return ERROR(KEY_NOT_FOUND, fmt::format("no network plugin found for name [{}]", _key));
+        }
+
+        _value = plugin_itr_->second;
+        return SUCCESS();
+
+    } // resolve
+
+    // =-=-=-=-=-=-=-
+    // deprecated - given a type, load up a network plugin
     error network_manager::init_from_type(
         const int&         _proc_type,
         const std::string& _type,
         const std::string& _key,
         const std::string& _inst,
         const std::string& _ctx,
-        network_ptr&       _net ) {
+        network_ptr&       _net)
+    {
+        if (_key != _inst) {
+            return ERROR(SYS_INVALID_INPUT_PARAM, "key and instance name must match");
+        }
```

Comment:
This function is deprecated in favor of the new `get_plugin` and will be removed at some point. I've consolidated this function and `resolve` into the new function so that thread-safety can be handled in one place.

Every call to this function that I have been able to find have used identical values for not only `_key` and `_inst`, but `_type` as well. I've not been able to find a use case where a different value for `_key` seems useful, so the new function does away with `_key` entirely and uses `_inst` for the legacy lookup table. This is why I've added this error case to the deprecated function.

Reply:
Got it. I know some types use inheritance, but this one doesn't, so we're good.